hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDTableServiceClient.java

```diff
@@ -320,9 +320,11 @@ private void updateTableMetadata(HoodieTable table, HoodieCommitMetadata commitM
 
   @Override
   protected void preCommit(HoodieCommitMetadata metadata) {
-    // Create a Hoodie table after startTxn which encapsulated the commits and files visible.
-    // Important to create this after the lock to ensure the latest commits show up in the timeline without need for reload
-    HoodieTable table = createTable(config, hadoopConf);
-    resolveWriteConflict(table, metadata, this.pendingInflightAndRequestedInstants);
+    if (config.getWriteConcurrencyMode().supportsOptimisticConcurrencyControl()) {
+      // Create a Hoodie table after startTxn which encapsulated the commits and files visible.
+      // Important to create this after the lock to ensure the latest commits show up in the timeline without need for reload
+      HoodieTable table = createTable(config, hadoopConf);
+      resolveWriteConflict(table, metadata, this.pendingInflightAndRequestedInstants);
+    }
```

Comment:
Maybe extract this to Base class to avoid code duplication, and override in Java class (not Flink and Spark) ?